File: classes/parte-7-paradigmas/111-herencia-composicion-y-polimorfismo/implementaciones/typescript/main.ts
import { readFileSync } from "node:fs";

interface Animal { sonido(): string; }
class Perro implements Animal { sonido() { return "guau"; } }
class Gato implements Animal { sonido() { return "miau"; } }
class Vaca implements Animal { sonido() { return "muu"; } }

const tipo: string = readFileSync(0, "utf8").trim();
const animales: Record<string, Animal> = { perro: new Perro(), gato: new Gato(), vaca: new Vaca() };
console.log(`sonido=${animales[tipo].sonido()}`);

File: classes/parte-7-paradigmas/111-herencia-composicion-y-polimorfismo/implementaciones/rust/main.rs
use std::io::Read;

trait Animal {
    fn sonido(&self) -> &'static str;
}

struct Perro;
struct Gato;
struct Vaca;

impl Animal for Perro { fn sonido(&self) -> &'static str { "guau" } }
impl Animal for Gato { fn sonido(&self) -> &'static str { "miau" } }
impl Animal for Vaca { fn sonido(&self) -> &'static str { "muu" } }

fn main() {
    let mut s = String::new();
    std::io::stdin().read_to_string(&mut s).unwrap();
    let tipo = s.trim();
    let a: Box<dyn Animal> = match tipo {
        "perro" => Box::new(Perro),
        "gato" => Box::new(Gato),
        _ => Box::new(Vaca),
    };
    println!("sonido={}", a.sonido());
}
